File: Catalog core/src/test/java/cz/vhromada/catalog/repository/MovieRepositoryIntegrationTest.java
package cz.vhromada.catalog.repository;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

import java.util.ArrayList;
import java.util.List;

import javax.persistence.EntityManager;

import cz.vhromada.catalog.commons.CollectionUtils;
import cz.vhromada.catalog.commons.GenreUtils;
import cz.vhromada.catalog.commons.MediumUtils;
import cz.vhromada.catalog.commons.MovieUtils;
import cz.vhromada.catalog.entities.Movie;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.data.domain.Sort;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.annotation.Rollback;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
import org.springframework.transaction.annotation.Transactional;

/**
 * A class represents integration test for class {@link MovieRepository}.
 *
 * @author Vladimir Hromada
 */
@RunWith(SpringJUnit4ClassRunner.class)
@ContextConfiguration("classpath:testRepositoryContext.xml")
@Transactional
@Rollback
public class MovieRepositoryIntegrationTest {

    /**
     * Instance of {@link EntityManager}
     */
    @Autowired
    @Qualifier("containerManagedEntityManager")
    private EntityManager entityManager;

    /**
     * Instance of {@link MovieRepository}
     */
    @Autowired
    private MovieRepository movieRepository;

    /**
     * Test method for get movies.
     */
    @Test
    public void testGetMovies() {
        final List<Movie> movies = movieRepository.findAll(new Sort("position", "id"));

        MovieUtils.assertMoviesDeepEquals(MovieUtils.getMovies(), movies);

        assertEquals(MovieUtils.MOVIES_COUNT, MovieUtils.getMoviesCount(entityManager));
        assertEquals(MediumUtils.MEDIA_COUNT, MediumUtils.getMediaCount(entityManager));
    }

    /**
     * Test method for get movie.
     */
    @Test
    public void testGetMovie() {
        for (int i = 1; i <= MovieUtils.MOVIES_COUNT; i++) {
            final Movie movie = movieRepository.findOne(i);

            MovieUtils.assertMovieDeepEquals(MovieUtils.getMovie(i), movie);
        }

        assertNull(movieRepository.findOne(Integer.MAX_VALUE));

        assertEquals(MovieUtils.MOVIES_COUNT, MovieUtils.getMoviesCount(entityManager));
        assertEquals(MediumUtils.MEDIA_COUNT, MediumUtils.getMediaCount(entityManager));
    }

    /**
     * Test method for add movie.
     */
    @Test
    @DirtiesContext
    public void testAdd() {
        final Movie movie = MovieUtils.newMovie(null);
        movie.setMedia(CollectionUtils.newList(MediumUtils.newMedium(null)));
        movie.setGenres(CollectionUtils.newList(GenreUtils.getGenre(entityManager, 1)));

        movieRepository.saveAndFlush(movie);

        assertNotNull(movie.getId());
        assertEquals(MovieUtils.MOVIES_COUNT + 1, movie.getId().intValue());

        final Movie addedMovie = MovieUtils.getMovie(entityManager, MovieUtils.MOVIES_COUNT + 1);
        final Movie expectedAddedMovie = MovieUtils.newMovie(null);
        expectedAddedMovie.setId(MovieUtils.MOVIES_COUNT + 1);
        expectedAddedMovie.setMedia(CollectionUtils.newList(MediumUtils.newMedium(MediumUtils.MEDIA_COUNT + 1)));
        expectedAddedMovie.setGenres(CollectionUtils.newList(GenreUtils.getGenre(1)));
        MovieUtils.assertMovieDeepEquals(expectedAddedMovie, addedMovie);

        assertEquals(MovieUtils.MOVIES_COUNT + 1, MovieUtils.getMoviesCount(entityManager));
        assertEquals(MediumUtils.MEDIA_COUNT + 1, MediumUtils.getMediaCount(entityManager));
    }

    /**
     * Test method for update movie with no media change.
     */
    @Test
    public void testUpdate_NoMediaChange() {
        final Movie movie = MovieUtils.updateMovie(entityManager, 1);

        movieRepository.saveAndFlush(movie);

        final Movie updatedMovie = MovieUtils.getMovie(entityManager, 1);
        final Movie expectedUpdatedMovie = MovieUtils.getMovie(1);
        MovieUtils.updateMovie(expectedUpdatedMovie);
        expectedUpdatedMovie.setPosition(MovieUtils.POSITION);
        MovieUtils.assertMovieDeepEquals(expectedUpdatedMovie, updatedMovie);

        assertEquals(MovieUtils.MOVIES_COUNT, MovieUtils.getMoviesCount(entityManager));
        assertEquals(MediumUtils.MEDIA_COUNT, MediumUtils.getMediaCount(entityManager));
    }

    /**
     * Test method for update movie with added medium.
     */
    @Test
    @DirtiesContext
    public void testUpdate_AddedMedium() {
        final Movie movie = MovieUtils.updateMovie(entityManager, 1);
        movie.getMedia().add(MediumUtils.newMedium(null));

        movieRepository.saveAndFlush(movie);

        final Movie updatedMovie = MovieUtils.getMovie(entityManager, 1);
        final Movie expectedUpdatedMovie = MovieUtils.getMovie(1);
        MovieUtils.updateMovie(expectedUpdatedMovie);
        expectedUpdatedMovie.getMedia().add(MediumUtils.newMedium(MediumUtils.MEDIA_COUNT + 1));
        expectedUpdatedMovie.setPosition(MovieUtils.POSITION);
        MovieUtils.assertMovieDeepEquals(expectedUpdatedMovie, updatedMovie);

        assertEquals(MovieUtils.MOVIES_COUNT, MovieUtils.getMoviesCount(entityManager));
        assertEquals(MediumUtils.MEDIA_COUNT + 1, MediumUtils.getMediaCount(entityManager));
    }

    /**
     * Test method for update movie with removed medium.
     */
    @Test
    public void testUpdate_RemovedMedium() {
        final int mediaCount = MovieUtils.getMovie(1).getMedia().size();
        final Movie movie = MovieUtils.updateMovie(entityManager, 1);
        movie.getMedia().clear();

        movieRepository.saveAndFlush(movie);

        final Movie updatedMovie = MovieUtils.getMovie(entityManager, 1);
        final Movie expectedUpdatedMovie = MovieUtils.getMovie(1);
        MovieUtils.updateMovie(expectedUpdatedMovie);
        expectedUpdatedMovie.setMedia(new ArrayList<>());
        expectedUpdatedMovie.setPosition(MovieUtils.POSITION);
        MovieUtils.assertMovieDeepEquals(expectedUpdatedMovie, updatedMovie);

        assertEquals(MovieUtils.MOVIES_COUNT, MovieUtils.getMoviesCount(entityManager));
        assertEquals(MediumUtils.MEDIA_COUNT - mediaCount, MediumUtils.getMediaCount(entityManager));
    }

    /**
     * Test method for remove movie.
     */
    @Test
    public void testRemove() {
        final int mediaCount = MovieUtils.getMovie(1).getMedia().size();

        movieRepository.delete(MovieUtils.getMovie(entityManager, 1));

        assertNull(MovieUtils.getMovie(entityManager, 1));

        assertEquals(MovieUtils.MOVIES_COUNT - 1, MovieUtils.getMoviesCount(entityManager));
        assertEquals(MediumUtils.MEDIA_COUNT - mediaCount, MediumUtils.getMediaCount(entityManager));
    }

    /**
     * Test method for remove all movies.
     */
    @Test
    public void testRemoveAll() {
        movieRepository.deleteAll();

        assertEquals(0, MovieUtils.getMoviesCount(entityManager));
        assertEquals(0, MediumUtils.getMediaCount(entityManager));
    }

}

File: catalog-core/src/test/kotlin/com/github/vhromada/catalog/repository/MovieRepositoryIntegrationTest.kt
package com.github.vhromada.catalog.repository

import com.github.vhromada.catalog.CatalogTestConfiguration
import com.github.vhromada.catalog.utils.AuditUtils
import com.github.vhromada.catalog.utils.GenreUtils
import com.github.vhromada.catalog.utils.MediumUtils
import com.github.vhromada.catalog.utils.MovieUtils
import com.github.vhromada.catalog.utils.TestConstants
import com.github.vhromada.catalog.utils.fillAudit
import com.github.vhromada.catalog.utils.updated
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions.assertSoftly
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.annotation.Rollback
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.transaction.annotation.Transactional
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext

/**
 * A class represents integration test for class [MovieRepository].
 *
 * @author Vladimir Hromada
 */
@ExtendWith(SpringExtension::class)
@ContextConfiguration(classes = [CatalogTestConfiguration::class])
@Transactional
@Rollback
class MovieRepositoryIntegrationTest {

    /**
     * Instance of [EntityManager]
     */
    @PersistenceContext
    private lateinit var entityManager: EntityManager

    /**
     * Instance of [MovieRepository]
     */
    @Autowired
    private lateinit var repository: MovieRepository

    /**
     * Test method for get movies.
     */
    @Test
    fun getMovies() {
        val movies = repository.findAll()

        MovieUtils.assertDomainMoviesDeepEquals(expected = MovieUtils.getMovies(), actual = movies)

        assertSoftly {
            it.assertThat(MovieUtils.getMoviesCount(entityManager = entityManager)).isEqualTo(MovieUtils.MOVIES_COUNT)
            it.assertThat(MediumUtils.getMediaCount(entityManager = entityManager)).isEqualTo(MediumUtils.MEDIA_COUNT)
        }
    }

    /**
     * Test method for get movie.
     */
    @Test
    fun getMovie() {
        for (i in 1..MovieUtils.MOVIES_COUNT) {
            val movie = repository.findById(i).orElse(null)

            MovieUtils.assertMovieDeepEquals(expected = MovieUtils.getMovieDomain(index = i), actual = movie)
        }

        assertThat(repository.findById(Int.MAX_VALUE)).isNotPresent

        assertSoftly {
            it.assertThat(MovieUtils.getMoviesCount(entityManager = entityManager)).isEqualTo(MovieUtils.MOVIES_COUNT)
            it.assertThat(MediumUtils.getMediaCount(entityManager = entityManager)).isEqualTo(MediumUtils.MEDIA_COUNT)
        }
    }

    /**
     * Test method for add movie.
     */
    @Test
    fun add() {
        val movie = MovieUtils.newMovieDomain(id = null)
            .copy(position = MovieUtils.MOVIES_COUNT, genres = listOf(GenreUtils.getGenre(entityManager = entityManager, id = 1)!!))
        val expectedMovie = MovieUtils.newMovieDomain(id = MovieUtils.MOVIES_COUNT + 1)
            .copy(
                media = listOf(MediumUtils.newMediumDomain(id = MediumUtils.MEDIA_COUNT + 1).fillAudit(audit = AuditUtils.newAudit())),
                picture = null,
                genres = listOf(GenreUtils.getGenreDomain(index = 1))
            ).fillAudit(audit = AuditUtils.newAudit())

        repository.save(movie)

        assertSoftly {
            it.assertThat(movie.id).isEqualTo(MovieUtils.MOVIES_COUNT + 1)
            it.assertThat(movie.createdUser).isEqualTo(TestConstants.ACCOUNT.uuid!!)
            it.assertThat(movie.createdTime).isEqualTo(TestConstants.TIME)
            it.assertThat(movie.updatedUser).isEqualTo(TestConstants.ACCOUNT.uuid!!)
            it.assertThat(movie.updatedTime).isEqualTo(TestConstants.TIME)
        }

        val addedMovie = MovieUtils.getMovie(entityManager, MovieUtils.MOVIES_COUNT + 1)!!
        assertThat(addedMovie).isNotNull
        MovieUtils.assertMovieDeepEquals(expected = expectedMovie, actual = addedMovie)

        assertSoftly {
            it.assertThat(MovieUtils.getMoviesCount(entityManager = entityManager)).isEqualTo(MovieUtils.MOVIES_COUNT + 1)
            it.assertThat(MediumUtils.getMediaCount(entityManager = entityManager)).isEqualTo(MediumUtils.MEDIA_COUNT + 1)
        }
    }

    /**
     * Test method for update movie.
     */
    @Test
    fun update() {
        val movie = MovieUtils.updateMovie(entityManager = entityManager, id = 1)
        val expectedMovie = MovieUtils.getMovieDomain(index = 1)
            .updated()
            .copy(position = MovieUtils.POSITION)
            .fillAudit(audit = AuditUtils.updatedAudit())

        repository.saveAndFlush(movie)

        val updatedMovie = MovieUtils.getMovie(entityManager = entityManager, id = 1)
        assertThat(updatedMovie).isNotNull
        MovieUtils.assertMovieDeepEquals(expected = expectedMovie, actual = updatedMovie!!)

        assertSoftly {
            it.assertThat(MovieUtils.getMoviesCount(entityManager = entityManager)).isEqualTo(MovieUtils.MOVIES_COUNT)
            it.assertThat(MediumUtils.getMediaCount(entityManager = entityManager)).isEqualTo(MediumUtils.MEDIA_COUNT)
        }
    }

    /**
     * Test method for update movie with added medium.
     */
    @Test
    @DirtiesContext
    fun updateAddedMedium() {
        var movie = MovieUtils.updateMovie(entityManager = entityManager, id = 1)
        val media = movie.media.toMutableList()
        media.add(MediumUtils.newMediumDomain(id = null))
        movie = movie.copy(media = media)
            .fillAudit(audit = movie)
        var expectedMovie = MovieUtils.getMovieDomain(index = 1)
            .updated()
        val expectedMedia = expectedMovie.media.toMutableList()
        expectedMedia.add(MediumUtils.newMediumDomain(id = MediumUtils.MEDIA_COUNT + 1).fillAudit(audit = AuditUtils.newAudit()))
        expectedMovie = expectedMovie.copy(media = expectedMedia)
            .fillAudit(audit = AuditUtils.updatedAudit())

        repository.saveAndFlush(movie)

        val updatedMovie = MovieUtils.getMovie(entityManager, 1)
        assertThat(updatedMovie).isNotNull
        MovieUtils.assertMovieDeepEquals(expected = expectedMovie, actual = updatedMovie!!)

        assertSoftly {
            it.assertThat(MovieUtils.getMoviesCount(entityManager)).isEqualTo(MovieUtils.MOVIES_COUNT)
            it.assertThat(MediumUtils.getMediaCount(entityManager)).isEqualTo(MediumUtils.MEDIA_COUNT + 1)
        }
    }

    /**
     * Test method for update movie with removed medium.
     */
    @Test
    fun updateRemovedMedium() {
        var movie = MovieUtils.updateMovie(entityManager = entityManager, id = 1)
        movie = movie.copy(media = emptyList())
            .fillAudit(audit = movie)
        val expectedMovie = MovieUtils.getMovieDomain(index = 1)
            .updated()
            .copy(media = emptyList())
            .fillAudit(audit = AuditUtils.updatedAudit())

        repository.saveAndFlush(movie)

        val updatedMovie = MovieUtils.getMovie(entityManager, 1)
        assertThat(updatedMovie).isNotNull
        MovieUtils.assertMovieDeepEquals(expected = expectedMovie, actual = updatedMovie!!)

        assertSoftly {
            it.assertThat(MovieUtils.getMoviesCount(entityManager)).isEqualTo(MovieUtils.MOVIES_COUNT)
            it.assertThat(MediumUtils.getMediaCount(entityManager)).isEqualTo(MediumUtils.MEDIA_COUNT - MovieUtils.getMovieDomain(index = 1).media.size)
        }
    }

    /**
     * Test method for remove movie.
     */
    @Test
    fun remove() {
        repository.delete(MovieUtils.getMovie(entityManager = entityManager, id = 1)!!)

        assertThat(MovieUtils.getMovie(entityManager = entityManager, id = 1)).isNull()

        assertSoftly {
            it.assertThat(MovieUtils.getMoviesCount(entityManager = entityManager)).isEqualTo(MovieUtils.MOVIES_COUNT - 1)
            it.assertThat(MediumUtils.getMediaCount(entityManager = entityManager)).isEqualTo(MediumUtils.MEDIA_COUNT - MovieUtils.getMovieDomain(index = 1).media.size)
        }
    }

    /**
     * Test method for remove all movies.
     */
    @Test
    fun removeAll() {
        repository.deleteAll()

        assertSoftly {
            it.assertThat(MovieUtils.getMoviesCount(entityManager = entityManager)).isEqualTo(0)
            it.assertThat(MediumUtils.getMediaCount(entityManager = entityManager)).isEqualTo(0)
        }
    }

    /**
     * Test method for get movies for user.
     */
    @Test
    fun findByCreatedUser() {
        val movies = repository.findByCreatedUser(user = AuditUtils.getAudit().createdUser!!)

        MovieUtils.assertDomainMoviesDeepEquals(expected = MovieUtils.getMovies(), actual = movies)

        assertSoftly {
            it.assertThat(MovieUtils.getMoviesCount(entityManager = entityManager)).isEqualTo(MovieUtils.MOVIES_COUNT)
            it.assertThat(MediumUtils.getMediaCount(entityManager = entityManager)).isEqualTo(MediumUtils.MEDIA_COUNT)
        }
    }

    /**
     * Test method for get movie by id for user.
     */
    @Test
    fun findByIdAndCreatedUser() {
        val user = AuditUtils.getAudit().createdUser!!
        for (i in 1..MovieUtils.MOVIES_COUNT) {
            val author = repository.findByIdAndCreatedUser(id = i, user = user).orElse(null)

            MovieUtils.assertMovieDeepEquals(expected = MovieUtils.getMovieDomain(index = i), actual = author)
        }

        assertThat(repository.findByIdAndCreatedUser(id = Int.MAX_VALUE, user = user)).isNotPresent

        assertSoftly {
            it.assertThat(MovieUtils.getMoviesCount(entityManager = entityManager)).isEqualTo(MovieUtils.MOVIES_COUNT)
            it.assertThat(MediumUtils.getMediaCount(entityManager = entityManager)).isEqualTo(MediumUtils.MEDIA_COUNT)
        }
    }

}
